### PR TITLE
Add support for ballistic beams

### PR DESCRIPTION
**Pull request by Baezon**
What it says on the tin. Surprisingly finicky to get the ammo to decrement properly. It shouldn't really matter too much but I semi-arbitrarily chose 4 per second as the ammo consumption rate for beams. Could be exposed but its kinda weird, ammo consumption is normally never specified because its always implicitly 1 per shot, but just like energy its nicer to for beams to have a continuous consumption.

Also moves energy handling into the common beam handling instead of just type c ones.

### CODE REVIEW

**Comment by Goober5000** [comment]
I would either make the ammo configurable or set it to 1 per shot.  Someone might decide to make a "conventional" primary weapon that uses the beam code for its effects.  In that case there would need to be 1 ammo consumed per beam.

And actually, to amend that slightly, it should be 1 ammo per firing point, not per shot.

File: code/weapon/beam.cpp [hunk 1]
```diff
@@ -1177,6 +1167,39 @@ void beam_move_all_post()
 			}
 		}
 
+		// deal with ammo/energy for fighter beams
+		bool multi_ai = MULTIPLAYER_CLIENT && (moveup->objp != Player_obj);
+		bool cheating_player = Weapon_energy_cheat && (moveup->objp == Player_obj);
+		if (moveup->flags & BF_IS_FIGHTER_BEAM && !multi_ai && !cheating_player) {
```
Comment (JohnAFernandez) [inline code/weapon/beam.cpp:1173]: Just add a line here to help coders see the start of the if block, and I think we're good.

**Review by JohnAFernandez — CHANGES_REQUESTED** [review]
I agree with Goober, but you didn't set out to create it like that.  Let's create a feature request with that behavior, once this is merged.

I only see one other thing.

**Comment by Baezon** [comment]
Well, he's kinda right, but also adding ammo consumption is gonna be awkward, I've already starting doing it and it's like, should a function which returns if a bank is out of ammo return if the bank has more than 0 ammo or more than the ammo consumption of the weapon currently loaded into the bank? It certainly would increase the scope of the PR quite a bit.

**Comment by JohnAFernandez** [comment]
Right, let's leave it for another PR, making sure there's a feature request for it so we don't forget.  Just make the other change, and we should merge this.

**Comment by Baezon** [comment]
Ok, so I've brought it back to the regular 1 ammo per beam, just like regular weapons. Might want to leave this 'till tomorrow though, as we've already merged quite a few things today.

**Review by JohnAFernandez — APPROVED** [review]

**Comment by Baezon** [comment]
By 'tomorrow' I of course meant a month.

**Pull request merged**